x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvAssert.java

```diff
@@ -239,6 +258,9 @@ public static void assertData(
                 throw ae;
             }
         }
+        if (dataFailures.isEmpty() == false) {
+            fail("Data mismatch: " + dataFailures);
+        }
```

Comment:
Not sure what the goal is here however the `DataFailure` doesn't trigger the AssertError catch on line on line 234/253. And that's an issue.

Reply:
The DataFailure approach actually provides improved details, when compared with the AssertError catch. Both approaches are trying to give more context to a failure, to make it easier to understand more about what failed, and how to fix it. In the AssertError catch code, you log the current line where the failure occurred. This will still happen for other types of failures, aside from the individual cell failure, like extra columns. The DateFailure approach tries to collect all failures together, so you can see the complete set of failures and fix them all in a bunch, instead of one at a time (which is a huge hassle for some tests). So even if this does not trigger the logging, it contains more information than that logging, so it does not matter.